Dashboard: update python parser

proy30 opened the pull request:
A branch that was worked on primarily during the summer months.

In cleanup processes now..

- [X] Support a wider range of impactX examples
- [X] Add `test_lattice_stats` 
     - By adding this test file, we test both the lattice statistics and indirectly the updated python parser. The statistics will reflect what the parser inputs into the dashboard. 
- [X] Support sim.lattice.append()
- [X] Support sim.lattice.extend()
- [X] Support sim.lattice.reverse()


_Note: As we rely on regex for the parser, it is "breakable", depending on the type of input files structures it receives. However, it is important to note that any of the impactX examples inside of the dashboard is **completely** supported by the parser.

proy30 committed:
place function for lattice parser in own file

will need it's own file to handle all the capabilities

proy30 committed:
break parse_lattice_elements into smaller helper functions

proy30 committed:
refactor: parse_cleaned_lattice

make more readable and clearly layout its purpose

proy30 committed:
add: collect_lattice_operations

proy30 committed:
add: inline_element_variables

will need to handle variable references

proy30 committed:
add: flatten_variable_definition

not yet completed

proy30 committed:
rebase for updating parse_cleaned_lattice

proy30 committed:
update: parse_lattice_elements to utilize helper functions

proy30 committed:
update naming: flatten variable list def

proy30 committed:
add test: run/fodo.py

proy30 committed:
add test: cyclotron

proy30 committed:
add test: dogleg

proy30 committed:
change if/else block to 'match'

does not affect performance, simpler to read

proy30 committed:
change static class to instance-based

for 'self. calls

proy30 committed:
fix indentation for match

proy30 committed:
fix: parsing blocking_factor

we just need the numerical number and not the brackets

proy30 committed:
update debug print statement for flatten

proy30 committed:
improve: flatten_variable_list_definitions

proy30 committed:
update collect_lattice_operations

do a greedy search of the parenthesis to stop at the last ')' in the line.

We need this because sim.extend() can have a element.(..) stored inside, and stopped at the first ')' is premature

proy30 committed:
rename flatton function name

proy30 committed:
add tests: expanding_fft_lattice

proy30 committed:
update function name

replace_variables_to_elements -> replace_variables

proy30 committed:
add .reverse() functionality/chicane test

proy30 committed:
add tests: apochromatic & expanding_fft - lattice builds

proy30 committed:
add docstring to class

proy30 committed:
update typends

proy30 committed:
add: caching for saving recursive time

proy30 committed:
separately find used lattice variables

proy30 committed:
update: docstrings/comments

proy30 committed:
store file_content in lattice class' initialization

proy30 committed:
add docstring

proy30 committed:
cache content hash

proy30 committed:
fix parser - distribution funciton

proy30 committed:
delete wrongly added file from rebase

proy30 committed:
revise naming for lattice vars/class/functions

proy30 committed:
simplify

update how lattice element counts are stored

add lattice stats test

expand on testdata/examples

cleanup/fix

make imports relative

fix

add tests

simplify

simplify parse_list_inputs

more comments/docstring

simplify more

cleanup tests: only keep 1-2 hardcoded tests to validate python parser

proy30 committed:
support .revsese()

found the culprit - the built in .reverse() that was being used emitted a reversed copy into the lattice instead of mutating the variable itself, and caused an extra set of elements to appear. we needed the reversed copy to be the only thing pushed to the elements, not the reversed copy AND the non-reversed

proy30 committed:
fix/cleanup

modify test_python_import's lattice section testing

test for correct population of element names and a single element parameter's value

I believe this is sufficient enough for testing how the lattice populates when importing from a file

fix test_lattice_stats and cleanup

fix examples

show examples

cleanup

[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

improvements

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 reviewed (commented):
Thanks a lot for this PR!  When I tried to pull this and initialize dashboard I received the attached error--it appears to be related to the previous PR #876.  Did you encounter this at any point?

<img width="1208" height="371" alt="Dashboard_error_22Sept2025" src="https://github.com/user-attachments/assets/413083ed-ec10-4e91-a0c9-ada5f439cf3b" />

Note:  proy30 is currently working to address this, and this PR shouldn't be merged until then.

proy30 committed:
relocate get_impactx_dir function

proy30 committed:
duplicate get_impactx_root_dir in tests utils

this funciton is needed by both the source code and the testing files. unsure where a good single location for the function would be without any import issues.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 commented:
Hi @cemitch99 , would you please be able to try again? I was unable to replicate my environment as yours, so I am not fully positive if the solution is now fixed or not.

cemitch99 commented:
Hi @proy30 .  Thanks for the recent changes!  I tried again, using `impactx-dashboard`, and I see the same error.  A similar error appears if I launch within Jupyter.  There could be an issue that is unique to my environment. 

However, if I launch the dashboard using `python -m dashboard` within `src/python/impactx`, then everything works perfectly.

proy30 commented:
@cemitch99 Thank you for letting me know. I have to investigate the issue further and will update again.

ax3l commented:
Fix in #1179

ax3l commented on src/python/impactx/dashboard/Input/utils.py:
```diff
@@ -139,3 +140,18 @@ def set_state_to_numeric(state_name: str) -> None:
+        current_directory = Path(__file__).resolve()
+        root_dir = None
+
+        for parent_dir in current_directory.parents:
```
@proy30 don't you already have a method for this from #876? :)

Please merge `development` and check if this is duplicate now :)

ax3l replied:
Ping @proy30 don't you already have a method for this from https://github.com/BLAST-ImpactX/impactx/pull/876? :)

proy30 replied:
@ax3l  Thank you for the ping. Yes, there is a duplicate function in both this PR (#1140) and #876.

At first, this was done intentionally when added in because (trying to recall correctly) it was difficult to have the pytests (from the directory `impactx/tests/python/dashboard/utils.py` call to the function in directory `impactx/src/python/impactx/dashboard/Input/utils.py`.

Any suggestions for where the function `get_impactx_root_dir` should be placed? It needs to be accessed by both the dashboard and the dashboard tests. Unsure if, in the future, the functionality could be used elsewhere where the root impactx directory is needed.

ax3l commented on src/python/impactx/dashboard/Input/utils.py:
```diff
@@ -139,3 +140,18 @@ def set_state_to_numeric(state_name: str) -> None:
+
+        for parent_dir in current_directory.parents:
+            if parent_dir.name == "impactx" and (parent_dir / ".git").is_dir():
+                root_dir = parent_dir  # keep going until we reach the highest match
```
@proy30 it looks to me like you are quite often assuming that ImpactX is used from its source tree. This is, outside of development, not the case.

We install ImpactX like this:
```bash
cmake -S . -B build -DImpactX_PYTHON=ON
cmake --build build -j 4 --target pip_install
```

which copies things outside of the source tree. Thus, looking for a `.git/` directory is not a logic that will work when ImpactX was installed/deployed/is used in production.

proy30 replied:
@ax3l  Thank you for pointing this out. Will add a commit to work well in both development and non-development

proy30 replied:
Will utilise the old path and logic from `get_impactx_path()` that you mention in the comment below.

<img width="815" height="417" alt="image" src="https://github.com/user-attachments/assets/eb5afce7-dc2a-4687-a916-47155a0e4f5a" />

ax3l replied:
Sounds good! :)

ax3l commented on src/python/impactx/dashboard/Toolbar/examples_loader/loader.py:
```diff
@@ -64,7 +64,7 @@ def _get_example_content(file_name: str) -> dict:
         Retrieve the selected ImpactX example file and populate the UI with its values.
         """
 
-        impactx_directory = DashboardExamplesLoader.get_impactx_path()
```
Keep using the old path and logic from `get_impactx_path()`, I fixed the logic in #1179

ax3l commented on tests/python/dashboard/utils.py:
```diff
@@ -19,6 +19,22 @@
+    for parent_dir in current_directory.parents:
+        if parent_dir.name == "impactx" and (parent_dir / ".git").is_dir():
+            root_dir = parent_dir  # keep going until we reach the highest match
+    return root_dir
```
See above, please don't add this `.git/` path based logic.

It should work if you rely on the one I fixed in #1179

ax3l commented:
@proy30 let me know when you are ready for me to review the updates discussed above :) :pray:

ax3l commented:
@proy30 let me know when you are ready for me to review the updates discussed above :) 🙏

proy30 committed:
remove duplicate method

both methods accomplish same directory

proy30 committed:
remove duplicate calling in tests

proy30 commented:
@ax3l Just now seeing the ping, this is completed now once tests pass 🙏

ax3l commented on src/python/impactx/dashboard/Toolbar/examples_loader/loader.py:
```diff
@@ -83,6 +83,7 @@ def load_impactx_examples() -> None:
 
         impactx_directory = DashboardExamplesLoader.get_impactx_path()
         impactx_examples_directory = impactx_directory / "examples"
+        print(" the examples directory is ", impactx_examples_directory)
```
```suggestion

```

ax3l commented on tests/python/dashboard/conftest.py:
```diff
@@ -1,6 +1,11 @@
 import pytest
 from seleniumbase import SB
 
+sys.path.insert(0, str(pathlib.Path(__file__).resolve().parents[3] / "src" / "python"))
+
```
Leftover from coding assistant testing things?

ax3l committed:
Remove Print

ax3l reviewed (approved):
Thank you! :)
